File: solutions/algrithoms/K-Similar Strings/dfs.js
/**
 * @param {string} s1
 * @param {string} s2
 * @return {number}
 */
let res;
const kSimilarity = function(s1, s2) {
    const len = s1.length;
    res = len;
    // const visited = new Set(); // 深度优先不能用 visited
    dfs(s1, s2, 0, 0, len)
    return res;
};

const dfs = (cur, s2, index, level, len) => {
    if (res <= level || res <= level + heuristic(cur, s2)) {
        return;
    }
    if (cur === s2) {
        res = res > level ? level: res;
        return;
    }
    while (index < len && s2[index] === cur[index]) {
        index++;
    }
    for (let i = index + 1; i < len; i++) {
        if (cur[i] === s2[index]) {
            const next = swap(cur, i, index);
            dfs(next, s2, index + 1, level + 1, len);
        }
    }
}

const swap = (str, i, j) => {
    const arr = [...str];
    const temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
    return arr.join('');
}

const heuristic = (s1, s2) => {
    let count = 0;
    for (let i = 0; i < s1.length; i++) {
        if (s1[i] !== s2[i]) {
            count++;
        }
    }
    return Math.floor((count + 1) / 2);
}

File: src/solutions/algrithoms/Number of Islands/dfs.cc
class Solution {
public:
    // unordered_map<int, int> m;
    int numIslands(vector<vector<char>>& grid) {
        int h = grid.size();
        int w = grid[0].size();
        vector<vector<bool>> visited(h, vector<bool>(w));
        // for (int i = 0; i < grid.size(); i++) {
        //     for (int j = 0; j < grid[0].size(); j++) {
        //         if (grid[i][j] == '1') {
        //             visited[i][j] = true;
        //         }
        //     }
        // }
        int level = 0;
        for (int i = 0; i < h; i++) {
            for (int j = 0; j < w; j++) {
                if (grid[i][j] == '1' && !visited[i][j]) {
                    dfs(grid, visited, h, w, i, j);
                    level++;
                }
            }
        }
        return level;
    }

    void dfs(vector<vector<char>>& grid, vector<vector<bool>>& visited, int h, int w, int i, int j) {
        visited[i][j] = true;
        if (i > 0 && grid[i - 1][j] == '1' && !visited[i - 1][j]) {
            dfs(grid, visited, h, w, i - 1, j);
        }
        if (i < h - 1 && grid[i + 1][j] == '1' && !visited[i + 1][j]) {
            dfs(grid, visited, h, w, i + 1, j);
        }
        if (j > 0 && grid[i][j - 1] == '1' && !visited[i][j - 1]) {
            dfs(grid, visited, h, w, i, j - 1);
        }
        if (j < w - 1 && grid[i][j + 1] == '1' && !visited[i][j + 1]) {
            dfs(grid, visited, h, w, i, j + 1);
        }
        
    }
};